File: php/src/Session.php
<?php

declare(strict_types=1);

namespace HopTop\Xrr;

use HopTop\Xrr\Exception\CassetteMissException;

class Session
{
    public function __construct(
        private Mode $mode,
        private FileCassette $cassette
    ) {}

    /**
     * Execute one interaction according to session mode.
     *
     * record:      calls $do(), saves to cassette, returns result.
     * replay:      loads from cassette, returns deserialized resp; never calls $do().
     * passthrough: calls $do(), never touches cassette.
     *
     * @throws CassetteMissException on replay miss
     */
    public function record(AdapterInterface $adapter, mixed $req, callable $do): mixed
    {
        return match ($this->mode) {
            Mode::Record      => $this->doRecord($adapter, $req, $do),
            Mode::Replay      => $this->doReplay($adapter, $req),
            Mode::Passthrough => $do($req),
        };
    }

    private function doRecord(AdapterInterface $adapter, mixed $req, callable $do): mixed
    {
        $resp = $do($req);

        $fp = $adapter->fingerprint($req);
        $this->cassette->save(
            $adapter->getId(),
            $fp,
            $adapter->serializeReq($req),
            $adapter->serializeResp($resp)
        );

        return $resp;
    }

    private function doReplay(AdapterInterface $adapter, mixed $req): mixed
    {
        $fp   = $adapter->fingerprint($req);
        $data = $this->cassette->load($adapter->getId(), $fp);

        return $adapter->deserializeResp($data['resp']);
    }
}

File: rs/src/session.rs
use crate::{cassette::FileCassette, error::XrrError, Adapter};

pub enum Mode {
    Record,
    Replay,
    Passthrough,
}

pub struct Session {
    mode: Mode,
    cassette: FileCassette,
}

impl Session {
    pub fn new(mode: Mode, cassette: FileCassette) -> Self {
        Self { mode, cassette }
    }

    pub fn record<A: Adapter>(
        &self,
        adapter: &A,
        req: &A::Req,
        do_: impl FnOnce() -> Result<A::Resp, XrrError>,
    ) -> Result<A::Resp, XrrError> {
        match self.mode {
            Mode::Record => {
                let resp = do_()?;
                let fp = adapter.fingerprint(req)?;
                self.cassette.save(adapter.id(), &fp, req, &resp)?;
                Ok(resp)
            }
            Mode::Replay => {
                let fp = adapter.fingerprint(req)?;
                let (_req, resp): (A::Req, A::Resp) =
                    self.cassette.load(adapter.id(), &fp)?;
                Ok(resp)
            }
            Mode::Passthrough => do_(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::adapters::exec::{ExecAdapter, ExecRequest, ExecResponse};
    use std::collections::HashMap;
    use tempfile::TempDir;

    fn req() -> ExecRequest {
        ExecRequest {
            argv: vec!["echo".into(), "hello".into()],
            stdin: "".into(),
            env: HashMap::new(),
        }
    }

    fn resp() -> ExecResponse {
        ExecResponse {
            stdout: "hello\n".into(),
            stderr: "".into(),
            exit_code: 0,
            duration_ms: 1,
        }
    }

    #[test]
    fn record_saves_and_returns() {
        let tmp = TempDir::new().unwrap();
        let cassette = FileCassette::new(tmp.path());
        let session = Session::new(Mode::Record, cassette);
        let adapter = ExecAdapter;
        let r = req();

        let result = session.record(&adapter, &r, || Ok(resp())).unwrap();
        assert_eq!(result.stdout, "hello\n");

        // Verify file was written.
        let fp = adapter.fingerprint(&r).unwrap();
        let path = tmp
            .path()
            .join(format!("exec-{}.req.yaml", fp));
        assert!(path.exists());
    }

    #[test]
    fn replay_loads_without_calling_do() {
        let tmp = TempDir::new().unwrap();
        let adapter = ExecAdapter;
        let r = req();
        let fp = adapter.fingerprint(&r).unwrap();

        // Pre-save cassette files.
        let cassette = FileCassette::new(tmp.path());
        cassette.save("exec", &fp, &r, &resp()).unwrap();

        let cassette2 = FileCassette::new(tmp.path());
        let session = Session::new(Mode::Replay, cassette2);

        let mut called = false;
        let result = session
            .record(&adapter, &r, || {
                called = true;
                Ok(resp())
            })
            .unwrap();

        assert!(!called, "do_ should not be called in replay mode");
        assert_eq!(result.stdout, "hello\n");
    }

    #[test]
    fn replay_miss_returns_cassette_miss() {
        let tmp = TempDir::new().unwrap();
        let cassette = FileCassette::new(tmp.path());
        let session = Session::new(Mode::Replay, cassette);
        let adapter = ExecAdapter;
        let r = req();

        let result = session.record(&adapter, &r, || Ok(resp()));
        assert!(matches!(result, Err(XrrError::CassetteMiss { .. })));
    }

    #[test]
    fn passthrough_calls_do_without_saving() {
        let tmp = TempDir::new().unwrap();
        let cassette = FileCassette::new(tmp.path());
        let session = Session::new(Mode::Passthrough, cassette);
        let adapter = ExecAdapter;
        let r = req();

        let mut called = false;
        let result = session.record(&adapter, &r, || {
            called = true;
            Ok(resp())
        }).unwrap();

        assert!(called);
        assert_eq!(result.exit_code, 0);

        // No files should exist.
        let entries: Vec<_> = std::fs::read_dir(tmp.path())
            .unwrap()
            .collect();
        assert!(entries.is_empty());
    }
}
